src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+            } while (currentState != knownState);
+
+            // Check for last reader
+            if (isLastReader)
+            {
+                // Check for waiting writers
+                if ((knownState & LockStates.WaitingWritersMask) != 0)
+                {
+                    Debug.Assert((_state & LockStates.WriterSignaled) != 0);
+                    Debug.Assert(writerEvent != null);
+                    writerEvent.Set();
+                }
+                // Check for waiting readers
+                else if ((knownState & LockStates.WaitingReadersMask) != 0)
+                {
+                    Debug.Assert((_state & LockStates.ReaderSignaled) != 0);
+                    Debug.Assert(readerEvent != null);
+                    readerEvent.Set();
+                }
+                // Check for the need to release events
+                else if (cacheEvents)
+                {
+                    ReleaseEvents();
+                }
+            }
+
+            Debug.Assert(lockEntry.IsEmpty);
+        }
+
+        public void ReleaseWriterLock()
+        {
+            if (_writerID != GetCurrentThreadID())
+            {
+                throw GetNotOwnerException();
+            }
+
+            Debug.Assert((_state & LockStates.ReadersMask) == 0);
+            Debug.Assert((_state & LockStates.Writer) != 0);
+            Debug.Assert(_writerLevel != 0);
+
+            // Check for nested release
+            --_writerLevel;
+            if (_writerLevel != 0)
+            {
+                return;
+            }
+
+            // Not a writer any more
+            _writerID = InvalidThreadID;
+            bool cacheEvents;
+            ManualResetEvent readerEvent = null;
+            AutoResetEvent writerEvent = null;
+            int currentState = _state;
+            int knownState;
+            do
+            {
+                cacheEvents = false;
+                knownState = currentState;
+                int modifyState = -LockStates.Writer;
+
+                if ((knownState & LockStates.WaitingReadersMask) != 0)
+                {
+                    try
+                    {
+                        readerEvent = GetOrCreateReaderEvent();
+                    }
+                    catch (SystemException)
+                    {
+                        Helpers.Sleep(100);
+                        currentState = _state;
+                        knownState = 0;
+                        Debug.Assert(currentState != knownState);
+                        continue;
+                    }
+                    modifyState += LockStates.ReaderSignaled;
+                }
+                else if ((knownState & LockStates.WaitingWritersMask) != 0)
+                {
+                    try
+                    {
+                        writerEvent = GetOrCreateWriterEvent();
+                    }
+                    catch (SystemException)
```

Comment:
In that case don't we run the risk of getting into an infinite loop here, e.g. in a low-resource situation?